utils/src/Vulkan/Utils/Requirements.hs

```diff
@@ -0,0 +1,554 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Vulkan.Utils.Requirements
+  ( -- * Instance requirements
+    checkInstanceRequirements
+  , -- * Device requirements
+    checkDeviceRequirements
+  , -- * Results
+    RequirementResult(..)
+  , Unsatisfied(..)
+  , requirementReport
+  , prettyRequirementResult
+  ) where
+
+import           Control.Arrow                  ( Arrow((***)) )
+import           Control.Monad
+import           Control.Monad.IO.Class
+import           Control.Monad.Trans.State
+import           Data.Bifunctor
+import           Data.ByteString                ( ByteString )
+import qualified Data.Dependent.Map            as DMap
+import           Data.Dependent.Map             ( DMap )
+import           Data.Dependent.Sum             ( DSum((:=>)) )
+import           Data.Foldable
+import           Data.Functor.Product           ( Product(..) )
+import qualified Data.HashMap.Strict           as Map
+import           Data.Kind                      ( Type )
+import           Data.List.Extra                ( nubOrd )
+import           Data.Proxy
+import           Data.Semigroup                 ( Endo(..) )
+import           Data.Traversable
+import           Data.Typeable                  ( eqT )
+import qualified Data.Vector                   as V
+import           Data.Vector                    ( Vector )
+import           Data.Word
+import           GHC.Base                       ( Proxy# )
+import           GHC.Exts                       ( proxy# )
+import           Type.Reflection
+import           Vulkan.CStruct                 ( FromCStruct
+                                                , ToCStruct
+                                                )
+import           Vulkan.CStruct.Extends
+import           Vulkan.Core10
+import           Vulkan.Core11.DeviceInitialization
+import           Vulkan.Core11.Promoted_From_VK_KHR_get_physical_device_properties2
+import           Vulkan.NamedType
+import           Vulkan.Requirement
+import           Vulkan.Version
+import           Vulkan.Zero                    ( Zero(..) )
+
+----------------------------------------------------------------
+-- * Instance Creation
+----------------------------------------------------------------
+
+checkInstanceRequirements
```

Comment:
This function doesn't check that the provided `InstanceCreateInfos es` is allowable, i.e. that all the extension structures are compatible with the instance. That's fine, I think it's reasonable to assume `InstanceCreateInfos es` is valid relative to what is returned by `enumerateInstanceExtensionProperties`, but at the moment (as far as I can see there), is no mechanism that would create a structure chain corresponding to a set of extensions that the user asks for (`makeInstanceCreateInfo` doesn't do anything about extension structures, it keeps the same chain). This was what the (now removed) `InstanceSetting` constructor was used for. I envisioned that a `SPIR-V` extension might well specifically ask for a field in an instance extension structure to be set, which is why I included it in `Requirement`.

Reply:
Ah, yeah, I was considering that it was much easier for the user to just `:&` on the extension structures (instead of carrying around all sorts of peek/poke/extends proofs in the InstanceRequirement GADT)

At the moment it's only these four:

https://github.com/expipiplus1/vulkan/blob/0a59ebc9ce87f6b60dca93afc122f1c78abe4e34/src/Vulkan/CStruct/Extends.hs#L872-L875